udev-117/extras/scsi_id/scsi_id.8
.TH SCSI_ID 8 "December 2003" "" "Linux Administrator's Manual"
.SH NAME
scsi_id \- retrieve and generate a unique SCSI identifier
.SH SYNOPSIS
.BI scsi_id
[\fIoptions\fP]
.SH "DESCRIPTION"
.B scsi_id
queries a SCSI device via the SCSI INQUIRY vital product data (VPD) page 0x80 or
0x83 and uses the resulting data to generate a value that is unique across
all SCSI devices that properly support page 0x80 or page 0x83.

If a result is generated it is sent to standard output, and the program
exits with a zero value. If no identifier is output, the program exits
with a non\-zero value.

\fBscsi_id\fP is primarily for use by other utilities such as \fBudev\fP
that require a unique SCSI identifier.

By default all devices are assume black listed, the \fB\-\-whitelisted\fP option must
be specified on the command line or in the config file for any useful
behaviour.

SCSI commands are sent directly to the device via the SG_IO ioctl
interface.

In order to generate unique values for either page 0x80 or page 0x83, the
serial numbers or world wide names are prefixed as follows.

Identifiers based on page 0x80 are prefixed by the character 'S', the SCSI
vendor, the SCSI product (model) and then the the serial number returned
by page 0x80. For example:

.sp
.nf
# scsi_id --page=0x80 --devpath=/block/sdg
SIBM     3542           1T05078453
.fi
.P

Identifiers based on page 0x83 are prefixed by the identifier type
followed by the page 0x83 identifier. For example, a device with a NAA
(Name Address Authority) type of 3 (also in this case the page 0x83
identifier starts with the NAA value of 6):

.sp
.nf
# /sbin/scsi_id --page=0x83 --devpath=/block/sdg
3600a0b80000b174b000000d63efc5c8c
.fi
.P

.SH OPTIONS
.TP
.BI \-\-fallback-to-sysfs
Always print information (model, vendor strings) about the device even
if it does not support VPD pages.
.TP
.BI \-\-ignore-sysfs
Ignore sysfs entries. Used for devices which are not represented as SCSI
devices, but understand SG_IO commands.
.TP
.BI \-\-blacklisted
The default behaviour \- treat the device as black listed, and do nothing
unless a white listed device is found in the scsi_id config\-file.
.TP
.BI \-\-device=\| device\^
Instead
of determining and creating a device node based on a sysfs dev
entry as done for the \fB\-s\fP, send SG_IO commands to
\fBdevice\fP, such as \fB/dev/sdc\fP.
This argument should also be used when invoked via udev to avoid problems
with creation of temporary files on not-yet writable directories.
.TP
.BI \-\-config=\| config\-file
Read configuration and black/white list entries from
.B config\-file
rather than the default
.B /etc/scsi_id.config
file.
.TP
.BI \-\-whitelisted
Treat the device as white listed. The \fB\-\-whitelisted\fP option must be specified
on the command line or in the scsi_id configuration file for
.B scsi_id
to generate any output.
.TP
.BI \-\-prefix-bus-id
Prefix the identification string with the  driver model (sysfs) bus id of
the SCSI device.
.TP
.BI \-\-page=\| 0x80 | 0x83 | pre-spc3-83
Use SCSI INQUIRY VPD page code 0x80, 0x83, or pre-spc3-83.
.sp
The default
behaviour is to query the available VPD pages, and use page 0x83 if found,
else page 0x80 if found, else nothing.
.sp
Page pre-spc3-83 should only be utilized for those scsi devices which
are not compliant with the SPC-2 or SPC-3 format for page 83.  While this
option is used for older model 4, 5, and 6 EMC Symmetrix devices, its
use with SPC-2 or SPC-3 compliant devices will fallback to the page 83
format supported by these devices.
.TP
.BI \-\-devpath=\| sysfs\-devpath
Generate an id for the
.B sysfs\-devpath.
The sysfs mount point must not be included.
.TP
.BI \-\-replace-whitespace
Reformat the output : replace all whitespaces by underscores.
.TP
.BI \-\-export
Export all data in KEY=<value> format used to import in other programs.
.TP
.BI \-\-verbose
Generate verbose debugging output.
.TP
.BI \-\-version
Display version number and exit.
.RE

.SH ENVIRONMENT

If the DEVPATH environment variable is set, scsi_id assumes it has been
invoked via udev (or some other hotplug program), and all errors or
warnings are sent using syslog.
.P

.SH "FILES"
.nf
.ft B
.ft
/etc/scsi_id.config                  configuration and black/white list entries
.fi
.LP
.SH "SEE ALSO"
.BR udev (7)
.SH AUTHORS
Developed by Patrick Mansfield <patmans@us.ibm.com> based on SCSI ID
source included in earlier linux 2.5 kernels, sg_utils source, and SCSI
specifications.


udev-117/ChangeLog
Summary of changes from v116 to v117
============================================

Dan Nicholson (2):
      extras: ignore built and generated files
      volume_id: create relative symlink when $(libdir) = $(usrlibdir)

Kay Sievers (15):
      usb_id: fail if vendor/product can not be retrieved
      rules: SUSE update
      firmware: do not print error if logger is missing
      volume_id: vfat - allow all possible sector sizes
      volume_id: LUKS - export version
      volume_id: ntfs - rely on valid master file table
      volume_id: bump version
      udevinfo: exclude "uevent" file from --attribute-walk
      udevadm: merge all udev tools into a single binary
      udevadm: accept command as option, like --help, --version
      udevadm: add info option --device-id-of-file=<file>
      Makefile: fix bogus version number than got committed
      udevadm: also return major==0 results for --device-id-of-file
      man: udevd.8 - remove udevcontrol section
      udevadm: control - allow command to be passed as option

MUNEDA Takahiro (1):
      man: fix udevadm.8 typo

Matthias Schwarzott (2):
      firmware: remove hardcoded path to logger
      rules: Gentoo update

VMiklos (1):
      rules: Frugalware update


Summary of changes from v115 to v116
============================================

Bryan Kadzban (1):
      rules: fix typos

Harald Hoyer (3):
      check line length after comment check and whitespace strip
      only install *.rules
      remove extra space from udevinfo symlink output

Kay Sievers (29):
      rules: fix two trivial typos
      rules: random and urandom are 0666
      rules: add REMOVE_CMD rule
      track "move" events to rename database and failed files
      rules: Gentoo update
      rules: add i2o driver rule
      man: recreate man pages
      volume_id: fix linux_raid metadata version 1.0 detection
      add $name substitution
      do not delete the device node with ignore_remove, but handle the event
      print warning for invalid TEST operations
      rules: do not delete /lib/udev/devices/ nodes on "remove"
      rules: remove broken nvram group assignment without any permission
      add /dev/rtc symlink if new rtc drivers are used
      increase WAIT_FOR_SYSFS timeout to 10 seconds
      rules: put bsd nodes in /dev/bsd/ directory
      path_id: fix for stacked class devices
      ignore device node names while restoring symlinks from the stack
      use SEQNUM in /dev/.udev/queue/ instead of devpath
      rules: add memstick module loading
      udevinfo: simplify symlink printing logic
      prevent wrong symlink creation if database disagress with current rules
      fix wrong variable used in logged string
      update README
      rule_generator: move all policy from write_net_rules to the rules file
      rules: call usb_id only for SUBSYSTEMS=="usb"
      rules: split out and fix persistent tape rules
      fix debug output string
      rule_generator: always match netif type in generated rule

Matthias Schwarzott (3):
      rules: Gentoo update
      rules: Gentoo update
      rules: Gentoo update

Michael Morony (1):
      set buffer size if strlcpy/strlcat indicate truncation

maximilian attems (1):
      correct includes in udev_selinux.c


Summary of changes from v114 to v115
============================================

Harald Hoyer (1):
      rules: fix typo in 80-drivers.rules

Kay Sievers (15):
      rules: add default rules
      rules: update SUSE rules
      rules: add packages rules
      rules: add ia64 rules
      rules: move md-raid rules to packages dir
      rules: run vol_id only for partitions
      rules: update Fedora rules
      edd_id: move persistent rules to its own file
      accept relative path for TEST
      rules: add iowarrior rule
      volume_id: fix sqashfs detection
      do not ignore dynamic rule if it is the last one in the list
      rule_generator: fix wrong DRIVERS!= logic
      rules: update Fedora
      Makefile: install default rules

Marco d'Itri (3):
      rules_generator: remove policy from write_cd_rules
      rules_generator: fix write_cd_rules when similar names exist in the root directory
      rules: Debian update


Summary of changes from v113 to v114
============================================

Hannes Reinecke (3):
      collect: extra to synchronize actions across events
      add $driver subtitution
      rules_generator: add S/390 persistent network support

Kay Sievers (24):
      rules_generator: remove executable flag from include file
      always unlink temporary file before creating new one
      rules: SUSE update
      volume_id: ext4 detection
      udevtrigger: allow to specify action string
      add option to RUN key to ignore the return value of the program
      use global udev_log variable instead of parameter in run_program
      add udev_rules_run() to handle RUN list
      move udev_utils_run.c into udev_rules.c
      rules: SUSE update
      name_list: rename loop_name -> name_loop
      handle dynamic rules created in /dev/.udev/rules.d/
      allow SYMLINK== match
      libvolume_id: use /usr/$libdir in pc file
      Makefile: add --as-needed flag to ld
      restore behavior of NAME==
      rules_generator: remove "installation" function
      udevtrigger: trigger "driver" events
      rules: update SUSE
      rules: Fedora update
      rules: add "do not edit" comment
      rules: Fedora update
      rules_generator: skip random MAC addresses
      write changed network interface names to the kernel log

Matthias Schwarzott (3):
      rules: Gentoo update
      fix inotify to work not only once
      rules: Gentoo update

Richard Hughes (1):
      Makefile: add "make dist" for nightly snapshots


Summary of changes from v112 to v113
============================================

David Zeuthen (1):
      vol_id: do not fail if unable to drop privileges

Kay Sievers (12):
      add missing ChangeLog

udev-117/extras/scsi_id/scsi_id.c
/*
 * scsi_id.c
 *
 * Main section of the scsi_id program
 *
 * Copyright (C) IBM Corp. 2003
 * Copyright (C) SUSE Linux Products GmbH, 2006
 *
 * Author:
 *	Patrick Mansfield<patmans@us.ibm.com>
 *
 *	This program is free software; you can redistribute it and/or modify it
 *	under the terms of the GNU General Public License as published by the
 *	Free Software Foundation version 2 of the License.
 */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <signal.h>
#include <fcntl.h>
#include <errno.h>
#include <string.h>
#include <syslog.h>
#include <stdarg.h>
#include <ctype.h>
#include <getopt.h>
#include <sys/stat.h>

#include "../../udev.h"
#include "scsi_id.h"
#include "scsi_id_version.h"

/* temporary names for mknod  */
#define TMP_DIR		"/dev"
#define TMP_PREFIX	"tmp-scsi"

static const struct option options[] = {
	{ "device", 1, NULL, 'd' },
	{ "config", 1, NULL, 'f' },
	{ "page", 1, NULL, 'p' },
	{ "devpath", 1, NULL, 's' },
	{ "fallback-to-sysfs", 0, NULL, 'a' },
	{ "blacklisted", 0, NULL, 'b' },
	{ "whitelisted", 0, NULL, 'g' },
	{ "prefix-bus-id", 0, NULL, 'i' },
	{ "replace-whitespace", 0, NULL, 'u' },
	{ "ignore-sysfs", 0, NULL, 'n' },
	{ "verbose", 0, NULL, 'v' },
	{ "version", 0, NULL, 'V' },
	{ "export", 0, NULL, 'x' },
	{ "help", 0, NULL, 'h' },
	{}
};

static const char short_options[] = "abd:f:ghinp:s:uvVx";
static const char dev_short_options[] = "bgp:";

static int all_good;
static int always_info;
static int dev_specified;
static int sys_specified;
static char config_file[MAX_PATH_LEN] = SCSI_ID_CONFIG_FILE;
static int display_bus_id;
static enum page_code default_page_code;
static int use_stderr;
static int debug;
static int hotplug_mode;
static int reformat_serial;
static int ignore_sysfs;
static int export;
static char vendor_str[64];
static char model_str[64];
static char revision_str[16];
static char type_str[16];

#ifdef USE_LOG
void log_message(int priority, const char *format, ...)
{
	va_list args;
	static int udev_log = -1;

	if (udev_log == -1) {
		const char *value;

		value = getenv("UDEV_LOG");
		if (value)
			udev_log = log_priority(value);
		else
			udev_log = LOG_ERR;
	}

	if (priority > udev_log)
		return;

	va_start(args, format);
	vsyslog(priority, format, args);
	va_end(args);
}
#endif

static void set_str(char *to, const char *from, size_t count)
{
	size_t i, j, len;

	/* strip trailing whitespace */
	len = strnlen(from, count);
	while (len && isspace(from[len-1]))
		len--;

	/* strip leading whitespace */
	i = 0;
	while (isspace(from[i]) && (i < len))
		i++;

	j = 0;
	while (i < len) {
		/* substitute multiple whitespace */
		if (isspace(from[i])) {
			while (isspace(from[i]))
				i++;
			to[j++] = '_';
		}
		/* skip chars */
		if (from[i] == '/') {
			i++;
			continue;
		}
		to[j++] = from[i++];
	}
	to[j] = '\0';
}

static void set_type(char *to, const char *from, size_t len)
{
	int type_num;
	char *eptr;
	char *type = "generic";

	type_num = strtoul(from, &eptr, 0);
	if (eptr != from) {
		switch (type_num) {
		case 0:
			type = "disk";
			break;
		case 1:
			type = "tape";
			break;
		case 4:
			type = "optical";
			break;
		case 5:
			type = "cd";
			break;
		case 7:
			type = "optical";
			break;
		case 0xe:
			type = "disk";
			break;
		case 0xf:
			type = "optical";
			break;
		default:
			break;
		}
	}
	strncpy(to, type, len);
	to[len-1] = '\0';
}

static int create_tmp_dev(const char *devpath, char *tmpdev, int dev_type)
{
	unsigned int maj, min;
	const char *attr;

	dbg("%s", devpath);
	attr = sysfs_attr_get_value(devpath, "dev");
	if (attr == NULL) {
		dbg("%s: could not get dev attribute: %s", devpath, strerror(errno));
		return -1;
	}

	dbg("dev value %s", attr);
	if (sscanf(attr, "%u:%u", &maj, &min) != 2) {
		err("%s: invalid dev major/minor", devpath);
		return -1;
	}

	snprintf(tmpdev, MAX_PATH_LEN, "%s/%s-maj%d-min%d-%u",
		 TMP_DIR, TMP_PREFIX, maj, min, getpid());

	dbg("tmpdev '%s'", tmpdev);
	if (mknod(tmpdev, 0600 | dev_type, makedev(maj, min))) {
		err("mknod failed: %s", strerror(errno));
		return -1;
	}
	return 0;
}

/*
 * get_value:
 *
 * buf points to an '=' followed by a quoted string ("foo") or a string ending
 * with a space or ','.
 *
 * Return a pointer to the NUL terminated string, returns NULL if no
 * matches.
 */
static char *get_value(char **buffer)
{
	static char *quote_string = "\"\n";
	static char *comma_string = ",\n";
	char *val;
	char *end;

	if (**buffer == '"') {
		/*
		 * skip leading quote, terminate when quote seen
		 */
		(*buffer)++;
		end = quote_string;
	} else {
		end = comma_string;
	}
	val = strsep(buffer, end);
	if (val && end == quote_string)
		/*
		 * skip trailing quote
		 */
		(*buffer)++;

	while (isspace(**buffer))
		(*buffer)++;

	return val;
}

static int argc_count(char *opts)
{
	int i = 0;
	while (*opts != '\0')
		if (*opts++ == ' ')
			i++;
	return i;
}

/*
 * get_file_options:
 *
 * If vendor == NULL, find a line in the config file with only "OPTIONS=";
 * if vendor and model are set find the first OPTIONS line in the config
 * file that matches. Set argc and argv to match the OPTIONS string.
 *
 * vendor and model can end in '\n'.
 */
static int get_file_options(const char *vendor, const char *model,
			    int *argc, char ***newargv)
{
	char *buffer;
	FILE *fd;
	char *buf;
	char *str1;
	char *vendor_in, *model_in, *options_in; /* read in from file */
	int lineno;
	int c;
	int retval = 0;

	dbg("vendor='%s'; model='%s'\n", vendor, model);
	fd = fopen(config_file, "r");
	if (fd == NULL) {
		dbg("can't open %s\n", config_file);
		if (errno == ENOENT) {
			return 1;
		} else {
			err("can't open %s: %s", config_file, strerror(errno));
			return -1;
		}
	}

	/*
	 * Allocate a buffer rather than put it on the stack so we can
	 * keep it around to parse any options (any allocated newargv
	 * points into this buffer for its strings).
	 */
	buffer = malloc(MAX_BUFFER_LEN);
	if (!buffer) {
		err("Can't allocate memory.");
		return -1;
	}

	*newargv = NULL;
	lineno = 0;
	while (1) {
		vendor_in = model_in = options_in = NULL;

		buf = fgets(buffer, MAX_BUFFER_LEN, fd);
		if (buf == NULL)
			break;
		lineno++;
		if (buf[strlen(buffer) - 1] != '\n') {
			info("Config file line %d too long.\n", lineno);
			break;
		}

		while (isspace(*buf))
			buf++;

		/* blank or all whitespace line */
		if (*buf == '\0')
			continue;

		/* comment line */
		if (*buf == '#')
			continue;

		dbg("lineno %d: '%s'\n", lineno, buf);
		str1 = strsep(&buf, "=");
		if (str1 && strcasecmp(str1, "VENDOR") == 0) {
			str1 = get_value(&buf);
			if (!str1) {
				retval = -1;
				break;
			}
			vendor_in = str1;

			str1 = strsep(&buf, "=");
			if (str1 && strcasecmp(str1, "MODEL") == 0) {
				str1 = get_value(&buf);
				if (!str1) {
					retval = -1;
					break;
				}
				model_in = str1;
				str1 = strsep(&buf, "=");
			}
		}

		if (str1 && strcasecmp(str1, "OPTIONS") == 0) {
			str1 = get_value(&buf);
			if (!str1) {
				retval = -1;
				break;
			}
			options_in = str1;
		}
		dbg("config file line %d:"
			" vendor '%s'; model '%s'; options '%s'\n",
			lineno, vendor_in, model_in, options_in);
		/*
		 * Only allow: [vendor=foo[,model=bar]]options=stuff
		 */
		if (!options_in || (!vendor_in && model_in)) {
			info("Error parsing config file line %d '%s'", lineno, buffer);
			retval = -1;
			break;
		}
		if (vendor == NULL) {
			if (vendor_in == NULL) {
				dbg("matched global option\n");
				break;
			}
		} else if ((vendor_in && strncmp(vendor, vendor_in,
						 strlen(vendor_in)) == 0) &&
			   (!model_in || (strncmp(model, model_in,
						  strlen(model_in)) == 0))) {
				/*
				 * Matched vendor and optionally model.
				 *
				 * Note: a short vendor_in or model_in can
				 * give a partial match (that is FOO
				 * matches FOOBAR).
				 */
				dbg("matched vendor/model\n");
				break;
		} else {
			dbg("no match\n");
		}
	}

	if (retval == 0) {
		if (vendor_in != NULL || model_in != NULL ||
		    options_in != NULL) {
			/*
			 * Something matched. Allocate newargv, and store
			 * values found in options_in.
			 */
			strcpy(buffer, options_in);
			c = argc_count(buffer) + 2;
			*newargv = calloc(c, sizeof(**newargv));
			if (!*newargv) {
				err("Can't allocate memory.");
				retval = -1;
			} else {
				*argc = c;
				c = 0;
				/*
				 * argv[0] at 0 is skipped by getopt, but
				 * store the buffer address there for
				 * later freeing
				 */
				(*newargv)[c] = buffer;
				for (c = 1; c < *argc; c++)
					(*newargv)[c] = strsep(&buffer, " \t");
			}
		} else {
			/* No matches  */
			retval = 1;
		}
	}
	if (retval != 0)
		free(buffer);
	fclose(fd);
	return retval;
}

static int set_options(int argc, char **argv, const char *short_opts,
		       char *target, char *maj_min_dev)
{
	int option;

	/*
	 * optind is a global extern used by getopt. Since we can call
	 * set_options twice (once for command line, and once for config
	 * file) we have to reset this back to 1.
	 */
	optind = 1;
	while (1) {
		option = getopt_long(argc, argv, short_opts, options, NULL);
		if (option == -1)
			break;

		if (optarg)
			dbg("option '%c' arg '%s'\n", option, optarg);
		else
			dbg("option '%c'\n", option);

		switch (option) {
		case 'a':
			always_info = 1;
			break;
		case 'b':
			all_good = 0;
			break;

		case 'd':
			dev_specified = 1;
			strncpy(maj_min_dev, optarg, MAX_PATH_LEN);
			maj_min_dev[MAX_PATH_LEN-1] = '\0';
			break;

		case 'e':
			use_stderr = 1;
			break;

		case 'f':
			strncpy(config_file, optarg, MAX_PATH_LEN);
			config_file[MAX_PATH_LEN-1] = '\0';
			break;

		case 'g':
			all_good = 1;
			break;

		case 'h':
			printf("Usage: scsi_id OPTIONS <device>\n"
			       "  --device               device node for SG_IO commands\n"
			       "  --devpath              sysfs devpath\n"
			       "  --config               location of config file\n"
			       "  --page                 SCSI page (0x80, 0x83, pre-spc3-83)\n"
			       "  --fallback-to-sysfs    print sysfs values if inquiry fails\n"
			       "  --ignore-sysfs         ignore sysfs entries\n"
			       "  --blacklisted          threat device as blacklisted\n"
			       "  --whitelisted          threat device as whitelisted\n"
			       "  --prefix-bus-id        prefix SCSI bus id\n"
			       "  --replace-whitespace   replace all whitespaces by underscores\n"
			       "  --verbose              verbose logging\n"
			       "  --version              print version\n"
			       "  --export               print values as environment keys\n"
			       "  --help                 print this help text\n\n");
			exit(0);

		case 'i':
			display_bus_id = 1;
			break;

		case 'p':
			if (strcmp(optarg, "0x80") == 0) {
				default_page_code = PAGE_80;
			} else if (strcmp(optarg, "0x83") == 0) {
				default_page_code = PAGE_83;
			} else if (strcmp(optarg, "pre-spc3-83") == 0) {
				default_page_code = PAGE_83_PRE_SPC3;
			} else {
				info("Unknown page code '%s'", optarg);
				return -1;
			}
			break;

		case 'n':
			ignore_sysfs = 1;
			break;

		case 's':
			sys_specified = 1;
			strncpy(target, optarg, MAX_PATH_LEN);
			target[MAX_PATH_LEN-1] = '\0';
			break;

		case 'u':
			reformat_serial = 1;
			break;

		case 'x':
			export = 1;
			break;

		case 'v':
			debug++;
			break;

		case 'V':
			printf("%s\n", SCSI_ID_VERSION);
			exit(0);
			break;

		default:
			exit(1);
		}
	}
	return 0;
}

static int per_dev_options(struct sysfs_device *dev_scsi, int *good_bad, int *page_code)
{
	int retval;
	int newargc;
	char **newargv = NULL;
	int option;

	*good_bad = all_good;
	*page_code = default_page_code;

	retval = get_file_options(vendor_str, model_str, &newargc, &newargv);

	optind = 1; /* reset this global extern */
	while (retval == 0) {
		option = getopt_long(newargc, newargv, dev_short_options, options, NULL);
		if (option == -1)
			break;

		if (optarg)
			dbg("option '%c' arg '%s'\n", option, optarg);
		else
			dbg("option '%c'\n", option);

		switch (option) {
		case 'b':
			*good_bad = 0;
			break;

		case 'g':
			*good_bad = 1;
			break;

		case 'p':
			if (strcmp(optarg, "0x80") == 0) {
				*page_code = PAGE_80;
			} else if (strcmp(optarg, "0x83") == 0) {
				*page_code = PAGE_83;
			} else if (strcmp(optarg, "pre-spc3-83") == 0) {
				*page_code = PAGE_83_PRE_SPC3;
			} else {
				info("Unknown page code '%s'", optarg);
				retval = -1;
			}
			break;

		default:
			info("Unknown or bad option '%c' (0x%x)", option, option);
			retval = -1;
			break;
		}
	}

	if (newargv) {
		free(newargv[0]);
		free(newargv);
	}
	return retval;
}

static int set_sysfs_values(struct sysfs_device *dev_scsi)
{
	const char *vendor, *model, *type;

	vendor = sysfs_attr_get_value(dev_scsi->devpath, "vendor");
	if (!vendor) {
		info("%s: cannot get vendor attribute", dev_scsi->devpath);
		return -1;
	}
	set_str(vendor_str, vendor, sizeof(vendor_str)-1);

	model = sysfs_attr_get_value(dev_scsi->devpath, "model");
	if (!model) {
		info("%s: cannot get model attribute\n", dev_scsi->devpath);
		return -1;
	}
	set_str(model_str, model, sizeof(model_str)-1);

	type = sysfs_attr_get_value(dev_scsi->devpath, "type");
	if (!type) {
		info("%s: cannot get type attribute", dev_scsi->devpath);
		return -1;
	}
	set_type(type_str, type, sizeof(type_str));

	type = sysfs_attr_get_value(dev_scsi->devpath, "rev");
	if (!type) {
		info("%s: cannot get type attribute\n", dev_scsi->devpath);
		return -1;
	}
	set_str(revision_str, type, sizeof(revision_str)-1);

	return 0;
}

static int set_inq_values(struct sysfs_device *dev_scsi, const char *path)
{
	int retval;
	char vendor[8], model[16], type[4], rev[4];

	retval = scsi_std_inquiry(dev_scsi, path, vendor, model, rev, type);
	if (retval)
	    return retval;

	set_str(vendor_str, vendor, 8);
	set_str(model_str, model, 16);
	set_type(type_str, type, sizeof(type_str) - 1);
	set_str(revision_str, rev, sizeof(revision_str) -1);

	return 0;
}

/*
 * format_serial: replace to whitespaces by underscores for calling
 * programs that use the serial for device naming (multipath, Suse
 * naming, etc...)
 */
static void format_serial(char *serial)
{
	char *p = serial, *q;

	q = p;
	while (*p != '\0') {
		if (isspace(*p)) {
			if (q > serial && q[-1] != '_') {
				*q = '_';
				q++;
			}
		} else {
			*q = *p;
			q++;
		}
		p++;
	}
	*q = '\0';
}

/*
 * scsi_id: try to get an id, if one is found, printf it to stdout.
 * returns a value passed to exit() - 0 if printed an id, else 1. This
 * could be expanded, for example, if we want to report a failure like no
 * memory etc. return 2, and return 1 for expected cases (like broken
 * device found) that do not print an id.
 */
static int scsi_id(const char *devpath, char *maj_min_dev)
{
	int retval;
	int dev_type = 0;
	struct sysfs_device *dev;
	struct sysfs_device *dev_scsi = NULL;
	int good_dev;
	int page_code;
	char serial[MAX_SERIAL_LEN];
	char serial_short[MAX_SERIAL_LEN];
	const char *bus_str = NULL;

	dbg("devpath %s\n", devpath);

	dev = sysfs_device_get(devpath);
	if (dev == NULL) {
		err("unable to access '%s'", devpath);
		return 1;
	}

	if (strcmp(dev->subsystem, "block") == 0)
		dev_type = S_IFBLK;
	else
		dev_type = S_IFCHR;

	/* mknod a temp dev to communicate with the device */
	if (!dev_specified && create_tmp_dev(dev->devpath, maj_min_dev, dev_type)) {
		dbg("create_tmp_dev failed\n");
		return 1;
	}

	if (!ignore_sysfs) {
		/* get scsi parent device */
		dev_scsi = sysfs_device_get_parent_with_subsystem(dev, "scsi");
		if (dev_scsi == NULL) {
			err("unable to access parent device of '%s'", devpath);
			return 1;
		}
		set_sysfs_values(dev_scsi);
		bus_str = "scsi";
	} else {
		dev_scsi = dev;
		set_inq_values(dev_scsi, maj_min_dev);
	}

	/* get per device (vendor + model) options from the config file */
	retval = per_dev_options(dev_scsi, &good_dev, &page_code);
	dbg("per dev options: good %d; page code 0x%x", good_dev, page_code);

	if (!good_dev) {
		retval = 1;
	} else if (scsi_get_serial(dev_scsi, maj_min_dev, page_code,
				   serial, serial_short, MAX_SERIAL_LEN)) {
		retval = always_info?0:1;
	} else {
		retval = 0;
	}
	if (!retval) {
		if (export) {
			char serial_str[MAX_SERIAL_LEN];

			printf("ID_VENDOR=%s\n", vendor_str);
			printf("ID_MODEL=%s\n", model_str);
			printf("ID_REVISION=%s\n", revision_str);
			set_str(serial_str, serial, sizeof(serial_str));
			printf("ID_SERIAL=%s\n", serial_str);
			set_str(serial_str, serial_short, sizeof(serial_str));
			printf("ID_SERIAL_SHORT=%s\n", serial_str);
			printf("ID_TYPE=%s\n", type_str);
			if (bus_str != NULL)
				printf("ID_BUS=%s\n", bus_str);
		} else {
			if (reformat_serial)
				format_serial(serial);
			if (display_bus_id)
				printf("%s: ", dev_scsi->kernel);
			printf("%s\n", serial);
		}
		dbg("%s\n", serial);
		retval = 0;
	}

	if (!dev_specified)
		unlink(maj_min_dev);

	return retval;
}

int main(int argc, char **argv)
{
	int retval = 0;
	char devpath[MAX_PATH_LEN];
	char maj_min_dev[MAX_PATH_LEN];
	int newargc;
	const char *env;
	char **newargv;

	logging_init("scsi_id");
	sysfs_init();
	dbg("argc is %d\n", argc);

	/* sysfs path can be overridden for testing */
	env = getenv("SYSFS_PATH");
	if (env) {
		strncpy(sysfs_path, env, sizeof(sysfs_path));
		sysfs_path[sizeof(sysfs_path)-1] = '\0';
	} else
		strcpy(sysfs_path, "/sys");

	env = getenv("DEVPATH");
	if (env) {
		hotplug_mode = 1;
		sys_specified = 1;
		strncpy(devpath, env, MAX_PATH_LEN);
		devpath[sizeof(devpath)-1] = '\0';
	}

	/*
	 * Get config file options.
	 */
	newargv = NULL;
	retval = get_file_options(NULL, NULL, &newargc, &newargv);
	if (retval < 0) {
		retval = 1;
		goto exit;
	}
	if (newargv && (retval == 0)) {
		if (set_options(newargc, newargv, short_options, devpath,
				maj_min_dev) < 0) {
			retval = 2;
			goto exit;
		}
		free(newargv);
	}

	/*
	 * Get command line options (overriding any config file or DEVPATH
	 * settings).
	 */
	if (set_options(argc, argv, short_options, devpath, maj_min_dev) < 0)
		exit(1);

	if (!sys_specified) {
		info("--devpath=<path> must be specified\n");
		retval = 1;
		goto exit;
	}

	retval = scsi_id(devpath, maj_min_dev);

exit:
	sysfs_cleanup();
	logging_close();
	return retval;
}
